index.ts

```diff
@@ -0,0 +1,10 @@
+import { IncomingMessage, ServerResponse } from 'http'
+
+type NextFunction = (err?: any) => void
```

Comment:
`err` should only be optional if you intend the user to check whether it was passed:

```ts
type NextFunction = (err?: any) => void
ienoopen(req, res, err => {
  if (err) {
    // do stuff
  }
  else {
    // err might not be passed! do different stuff!
  }
})
```

If you want to make it so (1) the user can assume that `err` is always passed and (2) the user can drop `err` if they don't handle errors, then you should `err` required in `NextFunction`.

```ts
type NextFunction = (err: any) => void
ienoopen(req, res, () => {
  // ok to ignore err, even if it's 'required'
}
```

That's because in the second example, it's required for *ienoopen* to pass `err`, but users can ignore any or all  callback parameters. In the first example, *ienoopen* is allowed to not pass `err`, but users *must* check whether it's undefined or not.

Reply:
I did this to match [Express's definition for `NextFunction`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express-serve-static-core/index.d.ts#L28-L31), but given that _I'm_ never going to call it with anything, could I do this instead?

```typescript
type NextFunction = () => void
```

This seemed to work in my testing but I don't want to break anyone's integration with Express because the types are incompatible. (And I could see a world where TypeScript says these types are incompatible.)

Reply:
Ah, I didn't know that Express does it that way.

Reply:
It _seems_ like that shorter definition still works. This seems to compile fine:

```typescript
type NextFunction = () => void
type ExpressNextFunction = (err?: any) => void

let foo: ExpressNextFunction
let bar: NextFunction

bar = () => { return }

foo = bar
```

If that's the case, am I safe to use the terser type definition?

Reply:
Also, I tried using Express in my tests with the shorter type definition and nothing complained, but it's possible there's something I'm missing.

Reply:
`bar = foo` works too, and will only happen in case `ienoopen` is itself passed to some express function. Yep, I think it's fine.